File: app/src/main/java/com/tichuguru/ScorecardFragment.java
package com.tichuguru;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import com.tichuguru.model.Game;
import com.tichuguru.model.Hand;
import com.tichuguru.model.Player;
import java.util.List;

public class ScorecardFragment extends Fragment {
    private TGViewModel viewModel;
    private RecyclerView scorecardList;
    private ScorecardAdapter adapter;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.scorecard, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        scorecardList = view.findViewById(R.id.scorecardList);
        scorecardList.setLayoutManager(new LinearLayoutManager(requireContext()));
        adapter = new ScorecardAdapter();
        scorecardList.setAdapter(adapter);
        view.findViewById(R.id.scorecardDelete).setOnClickListener(v -> onDeleteHand());
        viewModel = new ViewModelProvider(requireActivity()).get(TGViewModel.class);
        viewModel.getCurrentGame().observe(getViewLifecycleOwner(), game -> refreshDisplay());
    }

    @Override
    public void onHiddenChanged(boolean hidden) {
        super.onHiddenChanged(hidden);
        if (!hidden) refreshDisplay();
    }

    private void refreshDisplay() {
        Game game = TGApp.getGame();
        if (game == null) return;
        View view = requireView();
        List<Player> players = game.getPlayers();
        ((TextView) view.findViewById(R.id.scorecardName1)).setText(players.get(0).getName());
        ((TextView) view.findViewById(R.id.scorecardName2)).setText(players.get(1).getName());
        ((TextView) view.findViewById(R.id.scorecardName3)).setText(players.get(2).getName());
        ((TextView) view.findViewById(R.id.scorecardName4)).setText(players.get(3).getName());
        adapter.notifyDataSetChanged();
    }

    private void onDeleteHand() {
        new AlertDialog.Builder(requireContext())
            .setMessage("Are you sure?")
            .setPositiveButton("Yes", (dialog, which) -> {
                Game game = TGApp.getGame();
                game.removeHand(game.getHands().size() - 1);
                viewModel.notifyGameChanged();
            })
            .setNegativeButton("No", null)
            .show();
    }

    private static class ScorecardAdapter extends RecyclerView.Adapter<ScorecardAdapter.ViewHolder> {

        static class ViewHolder extends RecyclerView.ViewHolder {
            TextView score1, score2, total1, total2;
            TextView[] tichus = new TextView[4];

            ViewHolder(View v) {
                super(v);
                score1 = v.findViewById(R.id.scorecardHandScore1);
                score2 = v.findViewById(R.id.scorecardHandScore2);
                total1 = v.findViewById(R.id.scorecardTotalScore1);
                total2 = v.findViewById(R.id.scorecardTotalScore2);
                tichus[0] = v.findViewById(R.id.scorecardTichu1);
                tichus[1] = v.findViewById(R.id.scorecardTichu2);
                tichus[2] = v.findViewById(R.id.scorecardTichu3);
                tichus[3] = v.findViewById(R.id.scorecardTichu4);
            }
        }

        @NonNull
        @Override
        public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.scorecardrow, parent, false);
            return new ViewHolder(v);
        }

        @Override
        public void onBindViewHolder(ViewHolder holder, int position) {
            Game game = TGApp.getGame();
            Hand hand = game.getHands().get(position);
            int score = hand.getTotalScore1();
            holder.score1.setText((score >= 0 ? "+" : "") + score);
            int score2 = hand.getTotalScore2();
            holder.score2.setText((score2 >= 0 ? "+" : "") + score2);

            for (int i = 0; i < 4; i++) {
                TextView tv = holder.tichus[i];
                if (hand.isTichuFor(i)) tv.setText("T");
                else if (hand.isGrandTichuFor(i)) tv.setText("GT");
                else tv.setText("");
                tv.setTextColor(hand.outFirst() == i ? 0xFF00AA00 : -65536);
            }

            int s1 = 0, s2 = 0;
            for (int i = 0; i <= position; i++) {
                Hand h = game.getHands().get(i);
                s1 += h.getTotalScore1();
                s2 += h.getTotalScore2();
            }
            holder.total1.setText(String.valueOf(s1));
            holder.total2.setText(String.valueOf(s2));
        }

        @Override
        public int getItemCount() {
            Game game = TGApp.getGame();
            return game != null ? game.getHands().size() : 0;
        }
    }
}

File: app/src/main/kotlin/com/tichuguru/ScorecardFragment.kt
package com.tichuguru

import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.tichuguru.databinding.ScorecardBinding
import com.tichuguru.databinding.ScorecardrowBinding
import com.tichuguru.model.Game
import java.util.Locale

class ScorecardFragment : Fragment() {
    private lateinit var viewModel: TGViewModel
    private lateinit var binding: ScorecardBinding
    private lateinit var adapter: ScorecardAdapter

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        binding = ScorecardBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?,
    ) {
        super.onViewCreated(view, savedInstanceState)
        binding.scorecardList.layoutManager = LinearLayoutManager(requireContext())
        binding.scorecardDelete.setOnClickListener { onDeleteHand() }
        adapter = ScorecardAdapter()
        binding.scorecardList.adapter = adapter
        viewModel = ViewModelProvider(requireActivity())[TGViewModel::class.java]
        viewModel.getCurrentGame().observe(viewLifecycleOwner) { refreshDisplay() }
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        if (!hidden) refreshDisplay()
    }

    private fun refreshDisplay() {
        val game = viewModel.getCurrentGame().value ?: return
        val players = game.players
        binding.scorecardName1.text = players[0].name
        binding.scorecardName2.text = players[1].name
        binding.scorecardName3.text = players[2].name
        binding.scorecardName4.text = players[3].name
        adapter.game = game
    }

    private fun onDeleteHand() {
        confirmAction("Are you sure?") { viewModel.deleteLastHand() }
    }

    private class ScorecardAdapter : RecyclerView.Adapter<ScorecardAdapter.ViewHolder>() {
        private var cumTotals1: IntArray = intArrayOf()
        private var cumTotals2: IntArray = intArrayOf()

        var game: Game? = null
            @Suppress("NotifyDataSetChanged")
            set(value) {
                field = value
                value?.let { g ->
                    cumTotals1 = IntArray(g.hands.size)
                    cumTotals2 = IntArray(g.hands.size)
                    var t1 = 0
                    var t2 = 0
                    g.hands.forEachIndexed { i, hand ->
                        t1 += hand.totalScoreTeamOne(g.addOnFailure)
                        t2 += hand.totalScoreTeamTwo(g.addOnFailure)
                        cumTotals1[i] = t1
                        cumTotals2[i] = t2
                    }
                } ?: run {
                    cumTotals1 = intArrayOf()
                    cumTotals2 = intArrayOf()
                }
                notifyDataSetChanged()
            }

        class ViewHolder(val binding: ScorecardrowBinding) : RecyclerView.ViewHolder(binding.root) {
            val tichus =
                arrayOf(
                    binding.scorecardTichu1,
                    binding.scorecardTichu2,
                    binding.scorecardTichu3,
                    binding.scorecardTichu4
                )
        }

        override fun onCreateViewHolder(
            parent: ViewGroup,
            viewType: Int,
        ): ViewHolder = ViewHolder(ScorecardrowBinding.inflate(LayoutInflater.from(parent.context), parent, false))

        override fun onBindViewHolder(
            holder: ViewHolder,
            position: Int,
        ) {
            val game = checkNotNull(game) { "game not set" }
            val hand = game.hands[position]
            val s1 = hand.totalScoreTeamOne(game.addOnFailure)
            holder.binding.scorecardHandScore1.text = String.format(Locale.getDefault(), "%+d", s1)
            val s2 = hand.totalScoreTeamTwo(game.addOnFailure)
            holder.binding.scorecardHandScore2.text = String.format(Locale.getDefault(), "%+d", s2)

            holder.tichus.forEachIndexed { i, tv ->
                tv.text =
                    when {
                        hand.isTichuFor(i) -> "T"
                        hand.isGrandTichuFor(i) -> "GT"
                        else -> ""
                    }
                tv.setTextColor(if (hand.playerOutFirst == i) 0xFF00AA00.toInt() else Color.RED)
            }

            holder.binding.scorecardTotalScore1.text = String.format(Locale.getDefault(), "%d", cumTotals1[position])
            holder.binding.scorecardTotalScore2.text = String.format(Locale.getDefault(), "%d", cumTotals2[position])
        }

        override fun getItemCount() = game?.hands?.size ?: 0
    }
}
